### PR TITLE
fix #857 - fix on put when both @id and id are present

### LINKED ISSUE #857: PUT with embedded relation, no route matches "<int>", ItemNormalizer bug ?
Hello,

i'm facing a bug when submitting a PUT request with embedded relation :

data sample posted :

```
array:7 [
  "@context" => "/app_dev.php/contexts/Staking"
  "@id" => "/app_dev.php/stakings/10"
  "@type" => "Staking"
  "id" => 10
  "tournaments" => array:1 [
    0 => array:4 [
      "@id" => "/app_dev.php/staking_tournaments/3"
      "@type" => "StakingTournament"
      "id" => 3
      "tournament" => "/app_dev.php/tournaments/1"
    ]
  ]
  "name" => "Staking #111"
  "mode" => "simple"
]
```

api-platform returns me the following error :

```
{
    "@context":"\app_dev.php/contexts/Error",
    "@type":"Error",
    "hydra:title":"An error occurred",
    "hydra:description":"No route matches "3".", 
    ...
}
```

The error is caused by `src\Serializer\ItemNormalizer.php`, line 36, using `$data['id']` instead of `$data['@id']`. In my case, `$data['id']` is supposed to be an integer, the PK in database, not a resource URL.

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #857 
| License       | MIT
| Doc PR        | 

Add an non-regression test and modify the comportement of itemNormalizer

### CODE REVIEW

File: src/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -28,12 +28,12 @@
     public function denormalize($data, $class, $format = null, array $context = [])
     {
         // Avoid issues with proxies if we populated the object
-        if (isset($data['id']) && !isset($context['object_to_populate'])) {
+        if (isset($data['@id']) && !isset($context['object_to_populate'])) {
```
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: This snippet already exist in the JSON-LD normalizer: https://github.com/api-platform/core/blob/master/src/JsonLd/Serializer/ItemNormalizer.php#L93

`@id` is specific to JSON-LD, so it should not be here. However `id` can contain an IRI too (when not using JSON-LD). So it should be kept. I guess it's a priority problem, in https://github.com/api-platform/core/blob/master/src/JsonLd/Serializer/ItemNormalizer.php#L101 the call to the parent method should be done before.

**Comment by Simperfit** [comment]
@dunglas This is now working but i'm not sure about the fix, since I don't think it's in the same scope then the iriConverter is, clearly this is not an IRI, do you have any thing in mind that would be better ?

**Comment by Simperfit** [comment]
Again, maybe not the best fixes, but better than before since it's working and it does not break anything.

File: src/Serializer/ItemNormalizer.php [hunk 2]
```diff
@@ -32,8 +32,7 @@ public function denormalize($data, $class, $format = null, array $context = [])
             if (isset($context['api_allow_update']) && true !== $context['api_allow_update']) {
                 throw new InvalidArgumentException('Update is not allowed for this operation.');
             }
-
-            $context['object_to_populate'] = $this->iriConverter->getItemFromIri($data['id'], $context + ['fetch_data' => false]);
+            $context['object_to_populate'] = $this->iriConverter->getItemFromIri(is_int($data['id']) ? $this->iriConverter->getIriFromResourceClass($context['resource_class']).'/'.$data['id'] : $data['id'], $context + ['fetch_data' => false]);
```
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: id can be from other types (uuid are strings for instance).

File: tests/Fixtures/TestBundle/Entity/RelatedNormalizedDummy.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * Related to Normalized Dummy.
+ *
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ *
+ * @ApiResource(attributes={
+ *     "normalization_context"={"groups"={"related_output", "output"}},
+ *     "denormalization_context"={"groups"={"related_input", "input"}}
+ *
```
Comment (theofidry) [inline tests/Fixtures/TestBundle/Entity/RelatedNormalizedDummy.php]: useless extra line

File: tests/Fixtures/TestBundle/Entity/RelatedNormalizedDummy.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    public function __construct()
+    {
+        $this->customNormalizedDummy = new ArrayCollection();
+    }
+
+    /**
+     * @return int
+     */
+    public function getId()
```
Comment (theofidry) [inline tests/Fixtures/TestBundle/Entity/RelatedNormalizedDummy.php]: wouldn't a return typehint more suitable here and the other places in the same class?

File: features/main/custom_normalized.feature [hunk 5]
```diff
@@ -22,11 +22,66 @@ Feature: Using custom normalized entity
       "@context": "/contexts/CustomNormalizedDummy",
       "@id": "/custom_normalized_dummies/1",
       "@type": "CustomNormalizedDummy",
+      "id": 1,
       "name": "My Dummy",
       "alias": "My alias"
     }
     """
 
+  Scenario: Create a resource with relation
```
Comment (teohhanhui) [inline features/main/custom_normalized.feature]: Please fix the scenario summary (same for all below)...

**Comment by teohhanhui** [comment]
The fix is simple: We must not support embedded relations with `application/json`, because the semantics are undefined. We have had this discussion before on Slack, and I thought a decision was made...

/cc @api-platform/core-team

**Comment by dunglas** [comment]
@teohhanhui the issue here isn't with embedded relations but using raw ids instead of IRIs in raw JSON. Not sure we should support that too.

**Comment by teohhanhui** [comment]
What is the issue with deserializing using a raw id (or composite id)? I think it should be supported.

**Comment by Simperfit** [comment]
@teohhanhui updated

File: src/Serializer/ItemNormalizer.php [hunk 6]
```diff
@@ -33,7 +33,11 @@ public function denormalize($data, $class, $format = null, array $context = [])
                 throw new InvalidArgumentException('Update is not allowed for this operation.');
             }
 
-            $context['object_to_populate'] = $this->iriConverter->getItemFromIri($data['id'], $context + ['fetch_data' => false]);
+            try {
+                $context['object_to_populate'] = $this->iriConverter->getItemFromIri($data['id'], $context + ['fetch_data' => false]);
+            } catch (InvalidArgumentException $e) {
+                $context['object_to_populate'] = $this->iriConverter->getItemFromIri($this->iriConverter->getIriFromResourceClass($context['resource_class']).'/'.$data['id'], $context + ['fetch_data' => false]);
```
Comment (teohhanhui) [inline src/Serializer/ItemNormalizer.php]: How is this going to work with composite identifiers?

Comment (teohhanhui) [inline src/Serializer/ItemNormalizer.php]: And also this is hardcoded to use the property "id"? :cry:

Comment (Simperfit) [inline src/Serializer/ItemNormalizer.php]: @teohhanhui @dunglas do we want to support composite identifiers ?

**Comment by Simperfit** [comment]
ping @teohhanhui @dunglas modification done, WDYT ?

File: src/Serializer/ItemNormalizer.php [hunk 7]
```diff
@@ -33,7 +33,16 @@ public function denormalize($data, $class, $format = null, array $context = [])
                 throw new InvalidArgumentException('Update is not allowed for this operation.');
             }
 
-            $context['object_to_populate'] = $this->iriConverter->getItemFromIri($data['id'], $context + ['fetch_data' => false]);
+            try {
+                $context['object_to_populate'] = $this->iriConverter->getItemFromIri($data['id'], $context + ['fetch_data' => false]);
+            } catch (InvalidArgumentException $e) {
+                foreach ($this->propertyNameCollectionFactory->create($context['resource_class'], $context) as $propertyName) {
+                    if (true === $this->propertyMetadataFactory->create($context['resource_class'], $propertyName)->isIdentifier()) {
```
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: If we don't support yet composite identifiers, you can break after the first find. Or you can add support for composite identifiers :)

File: src/Serializer/ItemNormalizer.php [hunk 8]
```diff
@@ -33,7 +33,16 @@ public function denormalize($data, $class, $format = null, array $context = [])
...
+                        $identifier = $propertyName;
+                    }
+                }
+                $context['object_to_populate'] = $this->iriConverter->getItemFromIri($this->iriConverter->getIriFromResourceClass($context['resource_class']).'/'.$data[$identifier], $context + ['fetch_data' => false]);
```
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: I would use `sprintf` here for clarity.

File: src/Serializer/ItemNormalizer.php (hunk 7 above)
Comment (soyuka) [inline src/Serializer/ItemNormalizer.php]: +1 for the composite support :D

Comment (Simperfit) [inline src/Serializer/ItemNormalizer.php]: I think for now we should bug fix this and then add that feature ?

**Comment by dunglas** [comment]
@api-platform/core-team is everyone OK to merge this one too? I want to publish a bugfix release today.

**Comment by dunglas** [comment]
Thanks @Simperfit!

**Pull request merged**